Add possibility to use Spring Boot starters

herder opened the pull request:
This change means that you can utilize the Spring Boot starters (spring-cloud-gcp-*-starter, for example), to access the beans that they create. So you could use the Spring Boot pubsub support for example, by getting a handle to a PubsubTemplate via environment.resolve(PubsubTemplate.class)

One big limitation right now is that you can't define anything in your own app, so creating your own `@Configuration` classes for wiring your own beans in your app doesn't work, and creating Spring Data repositories won't work. The difficult part is trying to achieve this without the client code needing any changes - I think it means that we would need to change in the `StandaloneService` code so that it kicks off the autoconfiguration for the Service class that's provided to it. But I haven't tried that yet.

herder committed:
Add possibility to use Spring Boot starters

This makes it possible to use the Spring Boot starters in Apollo, by using `environment.resolve(klazz)`

It also makes the Lightbend configuration files available via @Value annotations, like for `.yml` and `.properties` files.

Now to make it possible to find the client app's own `@Configuration` classes, so we can autoconfigure there too.

Mandalorian007 commented on pom.xml:
```diff
@@ -51,6 +51,7 @@
         <junit.testkit.version>1.7.1</junit.testkit.version>
         <slf4j.version>1.7.30</slf4j.version>
         <hamcrest.version>2.2</hamcrest.version>
+        <spring.version>2.5.3</spring.version>
```
I think I would opt for labeling this as the spring-boot version over the Spring version

herder replied:
Done in fc3d1bda1b0a6b582cf16e46b01df1fe22eda4e9

Mandalorian007 commented on apollo-core/src/main/java/com/spotify/apollo/core/ServiceImpl.java:
```diff
@@ -207,12 +214,27 @@ InstanceImpl initInstance(
+           "spring.guice.autowireJIT=false",
+          //Allow overriding of bean definitions from different modules
+           "spring.main.allow-bean-definition-overriding=true")
+        .sources(GuiceSpringBridge.class)
```
Since you are adding this class as a source I am fairly certain you can leverage the @ComponentScan annotation on this class to enable picking up @Configuration, @Bean, and all variations of @Component

helenaAbergOstlund replied:
Thanks for the tip! This is working now after some struggles.

Mandalorian007 reviewed (commented):
Looks like a solid integration start I will have to pull it down and test it, but added a few small comments --mcf

helenaAbergOstlund committed:
Rename to spring.version to spring-boot.version

herder committed:
Put Spring integration behind feature flag

Can be activated via either env var `APOLLO_SPRING_ENABLED=true` or configuration option `apollo.spring.enabled=true`

helenaAbergOstlund committed:
Added @Configuration support for wiring your own beans

Mandalorian007 commented on apollo-core/src/main/java/com/spotify/apollo/core/ApolloInitializer.java:
```diff
@@ -0,0 +1,92 @@
+      configurableApplicationContext
+          .getBeanFactory()
+          .registerSingleton(findUniqueId(id, configurableApplicationContext), eachModule);
+      LOGGER.info(
```
This could get noisy. Would it make more sense to move this info to debug and maybe add a message that reports how many items were added to the context on info?

helenaAbergOstlund replied:
Figured it was ok since Apollo code also logs all loaded modules on INFO level.

Mandalorian007 replied:
All good then this will work it was just a thought it could add a bunch of logs

helenaAbergOstlund commented:
> One big limitation right now is that you can't define anything in your own app, so creating your own `@Configuration` classes for wiring your own beans in your app doesn't work, and creating Spring Data repositories won't work. The difficult part is trying to achieve this without the client code needing any changes - I think it means that we would need to change in the `StandaloneService` code so that it kicks off the autoconfiguration for the Service class that's provided to it. But I haven't tried that yet.

This has now been solved.

Mandalorian007 commented on apollo-core/src/main/java/com/spotify/apollo/core/GuiceSpringBridge.java:
```diff
@@ -0,0 +1,34 @@
+@EnableGuiceModules
+@SpringBootConfiguration
+@EnableAutoConfiguration
+@ComponentScan(basePackages = "com.spotify")
```
hmm so this will function fo any project that leverages a `com.spotify` package which places a bit of an interesting restriction on Apollo projects that would want to use Spring needing to have this package structure. Additionally, the usage of the `@EnableAutoConfiguration` at this package would treat the current package structure as the root package and would only configure the application context with things contained beneath it.

One idea that might actually make this a bit cleaner would be to have an @SpringBootApplication declared on the Main class within the Apollo Standalone (this annotation packages @Configuration, @EnableAutoConfiguration and @ComponentScan). Then since we need to register the main class as a source maybe we could leverage java reflections to search for any class annotated with @SpringBootApplication and register it.

helenaAbergOstlund replied:
Unfortunately I cannot get it to work without the `basePackages = "com.spotify"`. 

Your idea might work better but not something I will have time to try now as the hackweek is over.

herder replied:
I took a stab at fixing this a more generic way in 81cccfb
This makes the Spring bridge look for a `@SpringBootApplication`-annotated class to add as a source when initializing. 
See commit message for some questions, @helenaAbergOstlund @Mandalorian007

helenaAbergOstlund replied:
I like it!

> The main drawback is that the `main` class needs to be annotated with `@SpringBootApplication`, but on the other hand this makes the application a more by-the-book Spring Boot app.
> On the other hand this could be a feature: if we find an app annotated with `@SpringBootApplication` we auto-activate the Spring support, only skipping that if `apollo.spring.enabled` is explicitly set to `false`

Opting for the latter - assuming an app annotated with '@SpringBootApplication' will want to enable spring support - makes a lot of sense. I guess we need to change the code in that case so we always check the stack trace and scan for the annotation, and not only when the flag is enabled.

herder replied:
> Opting for the latter - assuming an app annotated with '@SpringBootApplication' will want to enable spring support - makes a lot of sense. I guess we need to change the code in that case so we always check the stack trace and scan for the annotation, and not only when the flag is enabled.

I agree: it makes more sense to try to follow the normal way of activating Spring as far as possible. One question is how to make this work with Fleet Management, though - it would require code changes besides just version bumping to get Spring support in all services. That might not be a real problem, though - I assume it's not too hard to create a fleet change that inspects the main class and adds the annotation if required. But this might be a problem for the future either way - we might cross that bridge when we need it :)

Mandalorian007 reviewed (commented):
Added some additional thoughts on Autowiring

helenaAbergOstlund committed:
Merge branch '1.x' into apollo-spring

helenaAbergOstlund committed:
Revert "Merge branch '1.x' into apollo-spring"

This reverts commit 695e65a64b44053e6e25e47f024759cf0fb49202, reversing
changes made to 49920b3be996b4f598f2db2c942bed28c946f7b5.

helenaAbergOstlund committed:
Manual merge from 1.x

helenaAbergOstlund committed:
Added test for spring option

helenaAbergOstlund committed:
Exclusions from missinglink-maven-plugin

helenaAbergOstlund commented:
Fixed conflicts with 1.x branch and the jdk8 build.

herder committed:
Add tests for Boot enabled/disabled

herder committed:
Find SpringBootApplication-annotated class

So that the component scan works as expected. This changes so that the Apollo Spring config tries to find an `@SpringBootApplication`-annotated class and add that as a source to the `SpringApplicationBuilder`.

This makes the component scan work more as expected in that it scans the packages under the found class, which is how Spring Boot normally works.

The main drawback is that the `main` class needs to be annotated with `@SpringBootApplication`, but on the other hand this makes the application a more by-the-book Spring Boot app.

On the other hand this could be a feature: if we find an app annotated with `@SpringBootApplication` we auto-activate the Spring support, only skipping that if `apollo.spring.enabled` is explicitly set to `false`

herder committed:
Make Spring deps optional

The only new dependency is the spring-guice library, with its spring-context dependency marked as excluded.

herder committed:
Only activate Spring support if found on classpath

If a class annotated with @SprinbBootApplication is found in the call stack at startup the Spring Boot support is enabled, unless the user explicitly has set apollo.spring.enabled to false.

herder committed:
Align dependencies

Jackson 2.12.3 -> 2.12.4, and added the spring-boot-dependencies bom as last import.

herder committed:
Adjust tests to new Spring activation strategy

herder committed:
Update header

herder committed:
Fix japicmp errors

The 0.6.1 version of the plugin failed with a ClassNotFoundError on optional dependencies, which the 0.15.3 version does not.

However, it instead failed by throwing an error about a new default method added to the com.spotify.apollo.core.Service interface, which is a feature that was introduced after 0.6.1.

The "overrideCompatibilityChangeParameters" section was added to fix this error.

herder committed:
Fix linking error

The Spring classes are an optional dependency - they are only used if they are detected on the classpath.

herder commented:
I changed the activation strategy to how I interpreted our discussion [here](https://github.com/spotify/apollo/pull/366#discussion_r702826082) so that Spring support is activated if one of these conditions is fulfilled:

- The main class is annotated with `@SpringBootApplication` and `apollo.spring.enabled` is not present or not set to `false`
- The `@SpringBootApplication` class is present on the classpath and `apollo.spring.enabled` is not set to `false`

https://github.com/spotify/apollo/pull/366/commits/47b51dedc77d76e3176e9600b7d2cff537cce6bb

Also, all the Spring dependencies are now optional except for `spring-guice`. To me it fits better with the 'just works' ideal that you only need to add the regular Spring dependencies if you want to make Apollo into a Spring app - you shouldn't have to care about how it is actually accomplished. And the classes are only loaded if Spring support is activated.

caesar-ralf commented:
This has moved to the internal repository
